File: ios/PteLiveIM/Sources/PteLiveIM/PteIMResponseCipher.swift
import CryptoKit
import Foundation

/** Implements the api-im P-256/A256GCM response envelope. */
internal enum PteIMResponseCipher {
  private static let context = Data("pte-live-api-response-v1".utf8)

  static func requestKey() -> P256.KeyAgreement.PrivateKey { P256.KeyAgreement.PrivateKey() }

  static func requestPublicKey(_ key: P256.KeyAgreement.PrivateKey) -> String {
    base64URL(key.publicKey.rawRepresentation)
  }

  static func decrypt(_ value: Data, with key: P256.KeyAgreement.PrivateKey) throws -> Data {
    let envelope = try JSONDecoder().decode(Envelope.self, from: value)
    guard envelope.version == 1, envelope.algorithm == "P-256/A256GCM" else { throw PteIMError.invalidResponse }
    let serverKey = try P256.KeyAgreement.PublicKey(rawRepresentation: try base64URLDecode(envelope.ephemeralPublicKey))
    let secret = try key.sharedSecretFromKeyAgreement(with: serverKey)
    let salt = try base64URLDecode(envelope.salt)
    let nonce = try base64URLDecode(envelope.nonce)
    guard salt.count == 32, nonce.count == 12 else { throw PteIMError.invalidResponse }
    let shared = secret.withUnsafeBytes { Data($0) }
    let symmetricKey = deriveKey(shared: shared, salt: salt, label: context)
    let combined = nonce + (try base64URLDecode(envelope.ciphertext))
    let box = try AES.GCM.SealedBox(combined: combined)
    return try AES.GCM.open(box, using: symmetricKey, authenticating: context)
  }

  static func deriveKey(shared: Data, salt: Data, label: Data) -> SymmetricKey {
    let material = shared + label
    let code = HMAC<SHA256>.authenticationCode(for: material, using: SymmetricKey(data: salt))
    return SymmetricKey(data: Data(code))
  }

  static func base64URL(_ value: Data) -> String {
    value.base64EncodedString().replacingOccurrences(of: "+", with: "-").replacingOccurrences(of: "/", with: "_").replacingOccurrences(of: "=", with: "")
  }

  static func base64URLDecode(_ value: String) throws -> Data {
    var base64 = value.replacingOccurrences(of: "-", with: "+").replacingOccurrences(of: "_", with: "/")
    base64 += String(repeating: "=", count: (4 - base64.count % 4) % 4)
    guard let data = Data(base64Encoded: base64) else { throw PteIMError.invalidResponse }
    return data
  }

  private struct Envelope: Decodable {
    let version: Int
    let algorithm: String
    let ephemeralPublicKey: String
    let salt: String
    let nonce: String
    let ciphertext: String
    enum CodingKeys: String, CodingKey {
      case version, algorithm, salt, nonce, ciphertext
      case ephemeralPublicKey = "ephemeral_public_key"
    }
  }
}

File: android/pte-im-sdk/src/main/java/com/ptelive/im/PteIMResponseCipher.kt
package com.ptelive.im

import android.util.Base64
import org.json.JSONObject
import java.math.BigInteger
import java.security.AlgorithmParameters
import java.security.KeyFactory
import java.security.KeyPair
import java.security.KeyPairGenerator
import java.security.PublicKey
import java.security.interfaces.ECPublicKey
import java.security.spec.ECGenParameterSpec
import java.security.spec.ECPoint
import java.security.spec.ECPublicKeySpec
import javax.crypto.Cipher
import javax.crypto.KeyAgreement
import javax.crypto.Mac
import javax.crypto.spec.GCMParameterSpec
import javax.crypto.spec.SecretKeySpec

/** Decrypts the short-lived P-256/A256GCM HTTP response envelopes required by api-im. */
internal object PteIMResponseCipher {
  private const val context = "pte-live-api-response-v1"

  fun createRequestKey(): KeyPair = KeyPairGenerator.getInstance("EC").apply {
    initialize(ECGenParameterSpec("secp256r1"))
  }.generateKeyPair()

  fun requestPublicKey(keyPair: KeyPair): String = encode(publicKeyBytes(keyPair.public as ECPublicKey))

  fun decrypt(envelope: JSONObject, keyPair: KeyPair): String {
    require(envelope.optInt("version") == 1 && envelope.optString("algorithm") == "P-256/A256GCM") { "unsupported API response encryption" }
    val shared = KeyAgreement.getInstance("ECDH").run {
      init(keyPair.private); doPhase(publicKey(decode(envelope.getString("ephemeral_public_key"))), true); generateSecret()
    }
    val salt = decode(envelope.getString("salt"))
    val nonce = decode(envelope.getString("nonce"))
    require(salt.size == 32 && nonce.size == 12) { "invalid API response encryption envelope" }
    val cipher = Cipher.getInstance("AES/GCM/NoPadding").apply {
      init(Cipher.DECRYPT_MODE, SecretKeySpec(deriveKey(shared, salt), "AES"), GCMParameterSpec(128, nonce))
      updateAAD(context.toByteArray(Charsets.UTF_8))
    }
    return cipher.doFinal(decode(envelope.getString("ciphertext"))).toString(Charsets.UTF_8)
  }

  fun publicKeyBytes(publicKey: ECPublicKey): ByteArray = byteArrayOf(4) +
    publicKey.w.affineX.fixed(32) + publicKey.w.affineY.fixed(32)

  fun publicKey(raw: ByteArray): PublicKey {
    require(raw.size == 65 && raw[0].toInt() == 4) { "invalid P-256 public key" }
    val parameters = AlgorithmParameters.getInstance("EC").apply { init(ECGenParameterSpec("secp256r1")) }
    val point = ECPoint(BigInteger(1, raw.copyOfRange(1, 33)), BigInteger(1, raw.copyOfRange(33, 65)))
    return KeyFactory.getInstance("EC").generatePublic(ECPublicKeySpec(point, parameters.getParameterSpec(java.security.spec.ECParameterSpec::class.java)))
  }

  fun deriveKey(shared: ByteArray, salt: ByteArray, label: String = context): ByteArray = Mac.getInstance("HmacSHA256").run {
    init(SecretKeySpec(salt, "HmacSHA256")); doFinal(shared + label.toByteArray(Charsets.UTF_8))
  }

  fun encode(value: ByteArray): String = Base64.encodeToString(value, Base64.URL_SAFE or Base64.NO_PADDING or Base64.NO_WRAP)
  fun decode(value: String): ByteArray = Base64.decode(value, Base64.URL_SAFE or Base64.NO_PADDING or Base64.NO_WRAP)
}

private fun BigInteger.fixed(size: Int): ByteArray {
  val raw = toByteArray().let { if (it.size > size && it[0].toInt() == 0) it.copyOfRange(1, it.size) else it }
  require(raw.size <= size) { "P-256 coordinate is too large" }
  return ByteArray(size - raw.size) + raw
}
